fix(web): properly update activityManager when browsing assets

robinbrisa opened the pull request:
## Description

On web, the ActivityManager widget does not work properly.
It seems to jump randomly when changing photos, sometimes showing the comment count for the whole album instead of the current photo.
The only way to reliably like a photo is by refreshing the page before clicking the like button.
This was caused by the album ActivityManager being initialized to the album even when viewing assets, as well as the asset ActivityManager not being properly initialized when switching assets.

Fixes #18524

First time contributing here and first time doing some front-end dev! Hope this helped.

## How Has This Been Tested?

- [x] Browse through a shared album, like a few photos. Refresh the page and check they are properly liked.
- [x] Like the shared album to make sure there is no regression.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

robinbrisa committed:
fix: properly update activityManager when browsing assets

danieldietzler commented on web/src/routes/(user)/albums/[albumId=id]/[[photos=photos]]/[[assetId=id]]/+page.svelte:
```diff
@@ -159,6 +159,10 @@
   };
 
   const updateComments = async () => {
+    if ($showAssetViewer) {
+      return;
+    }
```
Why is this necessary?

danieldietzler commented on web/src/lib/components/asset-viewer/asset-viewer.svelte:
```diff
@@ -140,6 +140,8 @@
 
   const updateComments = async () => {
     if (album) {
+      activityManager.reset();
+      activityManager.init(album.id, asset.id);
```
Resetting it for every update very much defeats the purpose of the store; caching counts when possible.

robinbrisa replied:
Hey Daniel

Thanks for your feedback :)

I'm very new to all of this so maybe I misunderstood a few things

From what I can tell, the activityManager looks like a singleton class that needs to be initialized every time the currently displayed asset changes. The reset() method just resets the internal state to zero and isn’t strictly necessary, since refreshActivities() will override the values anyway when it's called afterward.

It would definitely be useful to use it for caching counts, but as far as I can see, it doesn’t currently support that. We’d likely need to store a map internally keyed by album/asset pairs to do that properly (which I agree would be a worthwhile improvement).

The buttons in the activityManager trigger actions based on the album/asset pair that was last initialized. That’s actually what my changes were aiming to address:
- Avoiding the call to init() with a missing assetId when browsing assets (which should only happen in the album gallery view).
- Making sure init() is called when a new asset is displayed, so the buttons are correctly bound to that asset.


That said, I’m sure there are better or more idiomatic ways to do this. I still need to learn more about how things are typically done in TypeScript and SvelteKit, since these are new technologies for me.

danieldietzler reviewed (dismissed):
Hey, thanks! As is this isn't a proper fix for this issue. What you'd want is find the place where we don't reset the manager even though we should (stuff like unmounting, leaving a page, whatever), and only add a reset there.

robinbrisa committed:
Merge branch 'main' into fix/properly-update-activityManager

robinbrisa committed:
feat: added activityManager cache & reworked bugfix

robinbrisa committed:
fix: init activityManager on album page when shared

robinbrisa commented:
Okay, so I made some changes, partly according to your feedback:
- added cache for activityManager (based on album/asset pair), storing activities, comment count, like count and current user like status. Cache for last initialized album/asset pair is invalidated when adding/deleting activity.
- removed the updateComments methods. Now replaced by simply calling init(), which in addition to properly initializing the buttons for further actions also now calls refreshActivities (which was updateComments doing), using cache if necessary. Still calling this in $effect in +page.svelte for albums and the $asset-viewer component.
- hide activityManager widget after init() until refreshActivities() has finished running

It seems to be working fine to me and fixes the initial issue.

Only remaining issue I think is that it seems to make extra xhr queries for some reason, but that behavior was already happening before. However the addition of cache now prevents the queries from being dispatched. But it might be worth investigating

danieldietzler reviewed (commented):
Thank you very much! The code looks good to me, will have to test it still though :)

robinbrisa committed:
fix: invalidate album cache key after activity change on an asset

robinbrisa committed:
fix: wait until xhr is finished to display activityStatus widget (same behavior as AssetViewer)

robinbrisa commented:
> When an asset is commented on we should either manually increment the album counters in the activity manager, or we should make a new request to update the counts from the server.

Good point, this was an oversight, I just fixed it. Cheers

zackpollard commented:
Those changes look mostly good, the only thing is that it seems to be doubling up all of the requests after navigating back to the album. The activities calls and activities/statistics calls all get doubled up seemingly, see here. 
![image](https://github.com/user-attachments/assets/c5ae5368-bebd-49a0-ac1e-3c31b3a96732)

robinbrisa commented:
> Those changes look mostly good, the only thing is that it seems to be doubling up all of the requests after navigating back to the album. The activities calls and activities/statistics calls all get doubled up seemingly

Yeah, something causes the $effect hooks to be called multiple times. The cache key doesn't have the time to be created between calls so the XHR requests all go through.
I'm taking a look at this to finally fix all these duplicated calls. I just need a bit more knowledge on how svelte works :)

robinbrisa committed:
fix: prevent multiple assetManager inits

zackpollard commented:
I'm still seeing the same behavior as before, it's happening when I add a comment in an asset then return back to the album view

robinbrisa commented:
Seems fine to me.
Did the same exact steps and I'm getting 3 xhr queries per asset as expected (activities / liked / statistics)
After unvalidating the album cache (by liking or commenting an asset) you get these 3 queries called twice when visiting the album front page but it's no big deal as the 2nd call is cached (HTTP 304)
Viewing subsequent assets after that does the expect behavior, 3 xhr or 0 if cached.

zackpollard commented:
Yea, it was specifically navigating to the album page that I was talking about. It's fine, just unexpected behaviour. I am happy to merge this as-is if that's not something you want to fix right now :slightly_smiling_face:
